sympy/solvers/ode.py

```diff
@@ -6625,55 +6625,64 @@ def _linear_2eq_order1_type1(x, y, t, r, eq):
-    D = (r['a'] - r['d'])**2 + 4*r['b']*r['c']
-    if (r['a']*r['d'] - r['b']*r['c']) != 0:
-        if D > 0:
-            if r['b'].is_zero:
-                # tempting to use this in all cases, but does not guarantee linearly independent eigenvectors
-                gsol1 = C1*(l1 - r['d'] + r['b'])*exp(l1*t) + C2*(l2 - r['d'] + r['b'])*exp(l2*t)
-                gsol2 = C1*(l1 - r['a'] + r['c'])*exp(l1*t) + C2*(l2 - r['a'] + r['c'])*exp(l2*t)
-            else:
-                gsol1 = C1*r['b']*exp(l1*t) + C2*r['b']*exp(l2*t)
-                gsol2 = C1*(l1 - r['a'])*exp(l1*t) + C2*(l2 - r['a'])*exp(l2*t)
-        if D < 0:
-            sigma = re(l1)
-            if im(l1).is_positive:
-                beta = im(l1)
-            else:
-                beta = im(l2)
-            if r['b'].is_zero:
-                raise NotImplementedError('b == 0 case not implemented')
-            gsol1 = r['b']*exp(sigma*t)*(C1*sin(beta*t)+C2*cos(beta*t))
-            gsol2 = exp(sigma*t)*(((C1*(sigma-r['a'])-C2*beta)*sin(beta*t)+(C1*beta+(sigma-r['a'])*C2)*cos(beta*t)))
-        if D == 0:
-            if r['a']!=r['d']:
-                gsol1 = 2*r['b']*(C1 + C2/(r['a']-r['d'])+C2*t)*exp((r['a']+r['d'])*t/2)
-                gsol2 = ((r['d']-r['a'])*C1+C2+(r['d']-r['a'])*C2*t)*exp((r['a']+r['d'])*t/2)
-            if r['a']==r['d'] and r['a']!=0 and r['b']==0:
-                gsol1 = C1*exp(r['a']*t)
-                gsol2 = (r['c']*C1*t+C2)*exp(r['a']*t)
-            if r['a']==r['d'] and r['a']!=0 and r['c']==0:
-                gsol1 = (r['b']*C1*t+C2)*exp(r['a']*t)
-                gsol2 = C1*exp(r['a']*t)
-    elif (r['a']*r['d'] - r['b']*r['c']) == 0 and (r['a']**2+r['b']**2) > 0:
-        k = r['c']/r['a']
-        if r['a']+r['b']*k != 0:
-            gsol1 = r['b']*C1 + C2*exp((r['a']+r['b']*k)*t)
-            gsol2 = -r['a']*C1 + k*C2*exp((r['a']+r['b']*k)*t)
+    C1, C2 = get_numbered_constants(eq, num=2)
+    a, b, c, d = r['a'], r['b'], r['c'], r['d']
+    real_coeff = all(v.is_real for v in (a, b, c, d))
+    D = (a - d)**2 + 4*b*c
+    l1 = (a + d + sqrt(D))/2
+    l2 = (a + d - sqrt(D))/2
+    equal_roots = Eq(D, 0).simplify()
+    gsol1, gsol2 = [], []
+
+    # Solutions have exponential form if either D > 0 with real coefficients
+    # or D != 0 with complex coefficients. Eigenvalues are distinct.
+    # For each eigenvalue, pick an eigenvector, making sure we don't get (0, 0)
+    assumption = D > 0 if real_coeff else Not(equal_roots)
+    vector1 = Matrix((l1 - d, c)) if b == l1 - a == 0 else Matrix((b, l1 - a))
+    vector2 = Matrix((l2 - d, c)) if b == l2 - a == 0 else Matrix((b, l2 - a))
+    sol_vector = C1*exp(l1*t)*vector1 + C2*exp(l2*t)*vector2
+    gsol1.append((sol_vector[0], assumption))
+    gsol2.append((sol_vector[1], assumption))
+
+    # Solutions have trigonometric form for real coefficients with D < 0
+    # Both b and c are nonzero in this case, so (b, lam-a) is an eigenvector
```

Comment:
You refer to lambda as `lam` here but in the code use `l1` and `l2`. What is the exact relationship between those quantities?

Reply:
This is a statement applicable to both eigenvalues l1 or l2. I didn't want to type a single letter l as it's hard to read. I think this will be more clear if reworded as "so for any eigenvalue lam, (b, lam-a) is an eigenvector".